src/serialization/interval_log/mod.rs

```diff
@@ -0,0 +1,588 @@
+//!     .count();
+//!
+//! assert_eq!(3, count);
+//! ```
+//!
+//! Write a log.
+//!
+//! ```
+//! use std::str;
+//! use hdrsample;
+//! use hdrsample::serialization;
+//! use hdrsample::serialization::interval_log;
+//!
+//! let mut buf = Vec::new();
+//! let mut serializer = serialization::V2Serializer::new();
+//!
+//! let mut h = hdrsample::Histogram::<u64>::new_with_bounds(
+//!     1, u64::max_value(), 3).unwrap();
+//! h.record(12345).unwrap();
+//!
+//! // limit scope of mutable borrow of `buf`
+//! {
+//!     let mut header_writer = interval_log::IntervalLogHeaderWriter::new(
+//!         &mut buf, &mut serializer);
+//!     header_writer.write_comment("Comments are great").unwrap();
+//!     header_writer.write_start_time(123.456789).unwrap();
+//!
+//!     let mut log_writer = header_writer.into_log_writer();
+//!
+//!     log_writer.write_comment(
+//!         "You can have comments anywhere in the log").unwrap();
+//!
+//!     log_writer
+//!         .write_histogram(
+//!             &h,
+//!             1.234,
+//!             5.678,
+//!             interval_log::Tag::new("im-a-tag"),
+//!             1.0)
+//!         .unwrap();
+//! }
+//!
+//! // `buf` is now full of stuff; we check for the first line
+//! assert_eq!("#Comments are great\n", &str::from_utf8(&buf).unwrap()[0..20]);
+//! ```
+
+extern crate base64;
+
+use std::{fmt, io, ops, str};
+use std::fmt::Write;
+
+use nom::{double, line_ending, not_line_ending, IResult};
+
+use super::super::{Counter, Histogram};
+use super::Serializer;
+
+/// Write headers for an interval log.
+///
+/// This type only allows writing comments and headers. Once you're done writing those things, use
+/// `into_log_writer()` to convert this into an `IntervalLogWriter`.
+pub struct IntervalLogHeaderWriter<'a, 'b, W: 'a + io::Write, S: 'b + Serializer> {
+    internal_writer: InternalLogWriter<'a, 'b, W, S>,
+}
+
+impl<'a, 'b, W: 'a + io::Write, S: 'b + Serializer> IntervalLogHeaderWriter<'a, 'b, W, S> {
+    /// Create a new log writer that writes to `writer` and serializes histograms with `serializer`.
+    pub fn new(writer: &'a mut W, serializer: &'b mut S) -> IntervalLogHeaderWriter<'a, 'b, W, S> {
+        IntervalLogHeaderWriter {
+            internal_writer: InternalLogWriter {
+                writer,
+                serializer,
+                text_buf: String::new(),
+                serialize_buf: Vec::new(),
+            },
+        }
+    }
+
+    /// Add a comment line.
+    ///
+    /// If you do silly things like write a comment with a newline character, you'll end up with
+    /// an un-parseable log file. Don't do that.
+    pub fn write_comment(&mut self, s: &str) -> io::Result<()> {
+        self.internal_writer.write_comment(s)
+    }
+
+    /// Write a StartTime log line. See the module-level documentation for more info.
+    ///
+    /// This should only be called once to avoid creating a confusing interval log.
```

Comment:
Thinking about this some more, `IntervalLogHeaderWriter` really just seems like a builder to me. Maybe that's the way we should present it. That would allow us to only write things out when the user has committed to them, and then only write them once. Something like:
```rust
let mut ilb = IntervalLogBuilder::default().with_start_time(42);
ilb.add_comment("I love comments");
let mut logger = ilb.build(&mut writer, &mut serializer);
```
This has some other neat secondary effects such as headers being easy to re-use, fields that can only be set once only being written once, and more efficient writes (because they can be batch-written on `.build()`. We can bikeshed the exact name for `.build()`, but overall following the builder pattern as outlined in the [Rust API guidelines](https://aturon.github.io/ownership/builders.html) would be a good route to take.

Reply:
I like the builder approach.